Add "search_after" support to new termsEnum api

markharwood opened the pull request:
Adds an optional parameter to the `_terms_enum` request designed to allow paging. 
The last term from a previous result can be passed as the `search_after` parameter to a subsequent request, meaning only terms after the given term (but still matching the provided `string` prefix) are returned
Relates to https://github.com/elastic/elasticsearch/issues/72910

markharwood committed:
SearchAfter implementation but doesn’t work - it relies on field types returning a TermsEnum implementation that supports seekCeil() but we fell at the first hurdle with the keyword field not working - UnsupportedOperationException.
See https://github.com/elastic/elasticsearch/issues/72910 for discussion.

markharwood committed:
Moved searchAfter param to FieldType getTerms api so implementations can provide seek logic in setup without needing to return a TermsEnum that supports seekCeil() methods.

markharwood committed:
Unused imports

markharwood committed:
Line length fix

markharwood committed:
Added note to docs

markharwood committed:
Disable BWC tests until this is backported

markharwood committed:
Added fix and test for edge case - blank prefix used with search_after

markharwood committed:
Avoid toXContent for null searchAfter value

romseygeek commented on build.gradle:
```diff
@@ -213,9 +213,9 @@ tasks.register("verifyVersions") {
+boolean bwc_tests_enabled = false
 // place a PR link here when committing bwc changes:
-String bwc_tests_disabled_issue = ""
+String bwc_tests_disabled_issue = "https://github.com/elastic/elasticsearch/pull/72933"
```
Which BWC tests need to be disabled?

markharwood replied:
I think it was the REST yaml test for master -> 7.x that blew up because the search_after param is new.
Historically I've tried merging multiple PRs with temporary "if version = " checks while we do the backport dance and then clean them up but nearly always end up breaking builds so I opted for this big switch.

romseygeek replied:
I'm going to ping @pgomulka on this to get the best way of handling forward YAML tests - I think we can add the specific failing tests to a blocklist somewhere to avoid the big red button.

markharwood replied:
My current plan was:
1) Push this commit to master with all BWC turned off
2) Backport to 7.x but with all BWC flag turned back on
3) Push commit to master with BWC flag turned back on.


These are the less sweeping changes I _think_ I'd need. This is to modify the stream protocol for a 7.14 (ie currently unreleased) feature.
1. Make a change targeting master with 
  a) added BWC in/out stream version checks for on or after the current version (8.0) in master
  b) added skip to the yaml REST test for " - 7.99.99". 
3. Integrate the change into master.
4. Backport the change to 7.x, minus the stream version checks and rest skip header (instead keeping the existing REST tests skip in 7.x which already has  " - 7.13.99").
5. Integrate this change into 7.x.
6. Push one more change to master that removes the in/out BWC version checks and removes the rest skips  - (we only test BWC for 8.0 and last 7.x?).

@pgomulka can you verify?

romseygeek replied:
I hadn't spotted that this was a change to the stream API, I'd thought it was just the REST compat tests.  In which case, +1 to your original plan.

pgomulka replied:
I am probably not expert enough to recommend the best approach here. 
but I wonder if simply adding 
```
tasks.named("bwcTest").configure {

    systemProperty 'tests.rest.blacklist', [
    ]
}
```
would not work?
Could @elastic/es-delivery advice us here?

pgomulka replied:
@markharwood I think your plan is good and it should not affect the CI for others

markharwood replied:
> I think your plan is good and it should not affect the CI for others

Thanks for the feedback @pgomulka. Did you mean plan 1 (switching the global BWC flag)or plan 2 with adding the temporary version checking code (documented [above](https://github.com/elastic/elasticsearch/pull/72933#discussion_r632393931))?

markharwood replied:
In my second plan I don't think I need to add the version checks in the streaming code if we disable the rest tests (step 1b).
I've just tried pushing a change based on this approach to see if it builds OK.

romseygeek commented on server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java:
```diff
@@ -274,8 +277,33 @@ public TermsEnum getTerms(boolean caseInsensitive, String string, SearchExecutio
+        
+        // Initialises with a seek to a given term but excludes that term
+        // from results
+        public final class SearchAfterTermsEnum extends FilteredTermsEnum {
```
Does this need to be public?

markharwood replied:
I thought I might need it in FlattenedFieldMapper too but didn't in the end. I'll remove the public modifier.

romseygeek commented on server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java:
```diff
@@ -274,8 +277,33 @@ public TermsEnum getTerms(boolean caseInsensitive, String string, SearchExecutio
+            public SearchAfterTermsEnum(TermsEnum tenum, BytesRef termText) {
+                super(tenum);
+                afterRef = termText;
+                setInitialSeekTerm(termText);
```
I think you can simplify this a bit by just calling next() directly after `setInitialSeekTerm` here?

markharwood replied:
Will try it. This class was adapted from SingleTermsEnum which differs only in that the accept method returns NO if term is not equal.

markharwood replied:
Hmm. The test failed. The reason that led me to adding this class was that seek implementations tend to be "on or after" the provided term so you only want to call next if you're positioned _on_ the provided term. It doesn't work if the provided seek term is missing from the index (something we can expect when paging across multiple indices/shards).

romseygeek replied:
Ah yes, good point.  Maybe worth adding a comment in explaining this?

romseygeek commented on server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java:
```diff
@@ -440,10 +440,12 @@ public TextSearchInfo getTextSearchInfo() {
-    public TermsEnum getTerms(boolean caseInsensitive, String string, SearchExecutionContext queryShardContext) throws IOException {
+    public TermsEnum getTerms(boolean caseInsensitive, String string, SearchExecutionContext queryShardContext, BytesRef searchAfter)
+        throws IOException {
         return null;
```
We have a whole bunch of methods on MappedFieldType which deal only with text search, I wonder if we should move some of these into TextSearchInfo instead? One to put on the refactoring list...

romseygeek replied:
Given that the input string is a String, maybe the searchAfter parameter should be one too?

romseygeek reviewed (commented):
Thanks @markharwood! I left a few comments/suggestions.

markharwood committed:
Addressing review comments (thanks @romseygeek!)
searchAfter passed as string to field mappers, removed public modifier from class.

markharwood committed:
Visibility fix

romseygeek reviewed (approved):
LGTM! Thanks for the iters.  I'd like to get another comment on how to deal with BWC rest compat tests before we merge though, turning off all BWC tests makes me very nervous.

markharwood committed:
Added comments

markharwood committed:
Revert "Disable BWC tests until this is backported"

This reverts commit 81939b4265bb2e71c6f8c0e5fa22f2b1cfe5d64c.

markharwood committed:
More selective disabling of BWC tests while we handle the back port of search_after param

elasticmachine commented:
Pinging @elastic/es-search (Team:Search)
